### PR TITLE
Do data validation when undo.

### PR DESCRIPTION
**Pull request by ujjboy**
### Ⅰ. Describe what this PR did

- Add logic:
  - compare current data in the database with destination image, if they are not equal, it means have dirty data,
  - then compare current data in the database and the original image, if they are equal, also no need to execute undo
  - else  will throw an exception
- Add some test case

### Ⅱ. Does this pull request fix one issue?

Part of #1049

### CODE REVIEW

**Comment by ujjboy** [comment]
I will add more test case which uses h2 database later.

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java [hunk 1]
```diff
@@ -145,9 +164,115 @@ protected void undoPrepare(PreparedStatement undoPST, ArrayList<Field> undoValue
      * Data validation.
      *
      * @param conn the conn
-     * @throws SQLException the sql exception
+     * @return return true if data validation is ok and need continue
+     * @throws SQLException the sql exception such as has dirty data
      */
-    protected void dataValidation(Connection conn) throws SQLException {
+    protected boolean dataValidation(Connection conn) throws SQLException {
+
```
Comment (zhangthen) [inline rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java]: It ‘s better to add a switcher to close  the dirty data check manually, the switcher is open default.

File: rm-datasource/src/test/java/io/seata/rm/datasource/undo/UndoExecutorTest.java [hunk 2]
```diff
@@ -287,9 +294,12 @@ public void testDelete() {
...
-
+        MockConnection connection = new MockConnection();
+        AbstractUndoExecutor spy = Mockito.spy(executor);
         try {
-            executor.executeOn(new MockConnection());
+            // skip data validation
+            Mockito.doReturn(true).when(spy).dataValidation(connection);
+            spy.executeOn(connection);
         } catch (SQLException e) {
             e.printStackTrace();
```
Comment (leizhiyuan) [inline rm-datasource/src/test/java/io/seata/rm/datasource/undo/UndoExecutorTest.java]: don't use e.printTrace()

File: rm-datasource/src/test/java/io/seata/rm/datasource/undo/UndoExecutorTest.java [hunk 3]
```diff
@@ -217,9 +221,12 @@ public void testInsert() {
...
```
Comment (leizhiyuan) [inline rm-datasource/src/test/java/io/seata/rm/datasource/undo/UndoExecutorTest.java]: don't use e.printTrace()

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java (hunk 1 above)
Comment (ujjboy) [reply rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java]: OK, I will add it in the next standalone PR.

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java [hunk 4]
```diff
@@ -72,11 +91,11 @@ public SQLUndoLog getSqlUndoLog() {
      */
     public void executeOn(Connection conn) throws SQLException {
 
-        // no need undo if the before data snapshot is equivalent to the after data snapshot.
-        if (DataCompareUtils.isRecordsEquals(sqlUndoLog.getBeforeImage(), sqlUndoLog.getAfterImage())) {
+        // when the data validation is not ok 
+        if (!dataValidation(conn)) {
             return;
```
Comment (leizhiyuan) [inline rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java]: should we throw some exception explicitly for the code change later？

Comment (ujjboy) [reply rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java]: Sorry, I don’t get your point.

Comment (ujjboy) [reply rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java]: I add some log and change the method name for more explicitly.

Comment (leizhiyuan) [reply rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java]: as communicated offline

Comment (ujjboy) [reply rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java]: @leizhiyuan   done.

**Pull request merged**